src/test/bls_tests.cpp

```diff
@@ -51,4 +52,124 @@ BOOST_AUTO_TEST_CASE(bls_sig_tests)
     BOOST_CHECK(!sig2.VerifyInsecure(sk1.GetPublicKey(), msgHash2));
 }
 
+static BLSIdVector GetRandomBLSIds(size_t n)
+{
+    BLSIdVector v;
+    for (size_t i = 0; i < n; i++) {
+        v.emplace_back(GetRandHash());
+    }
+    return v;
+}
+
+std::vector<size_t> GetRandomElements(size_t m, size_t n)
+{
+    assert(m <= n);
+    std::vector<size_t> idxs;
+    for (size_t i = 0; i < n; i++) {
+        idxs.emplace_back(i);
+    }
+    std::random_shuffle(idxs.begin(), idxs.end());
+    return std::vector<size_t>(idxs.begin(), idxs.begin() + m);
+}
+
+struct Member
+{
+    CBLSId id;
+    BLSVerificationVectorPtr vecP;
+    BLSSecretKeyVector contributions;
+    CBLSSecretKey skShare;
+
+    Member(const CBLSId& _id): id(_id) {}
+};
+
+BOOST_AUTO_TEST_CASE(dkg)
+{
+    CBLSWorker worker;
+    const size_t N = 40;     // quorum size
+    const size_t M = 30;     // threshold
+    BLSSecretKeyVector allSkShares;
+
+    worker.Start();
+
+    // Create N Members and send contributions
+    const BLSIdVector& ids = GetRandomBLSIds(N);
+    std::vector<Member> quorum;
+    for (const auto& id : ids) {
+        Member m(id);
+        // Generate contributions
+        worker.GenerateContributions((int)M, ids, m.vecP, m.contributions);
+        BOOST_CHECK_EQUAL(m.vecP->size(), M);
+        BOOST_CHECK_EQUAL(m.contributions.size(), N);
+        // Verify contributions against verification vector
+        for (size_t j = 0; j < N; j++) {
+            BOOST_CHECK(worker.VerifyContributionShare(ids[j], m.vecP, m.contributions[j]));
+        }
+        quorum.emplace_back(std::move(m));
+    }
+
+    // Aggregate received contributions for each Member to produce key shares
+    for (size_t i = 0; i < N; i++) {
+        Member& m = quorum[i];
+        BLSSecretKeyVector rcvContributions;
+        for (size_t j = 0; j < N; j++) {
+            rcvContributions.emplace_back(quorum[j].contributions[i]);
+        }
+        m.skShare = worker.AggregateSecretKeys(rcvContributions);
```

Comment:
Can verify the validity of this sk share below with a:

```
// Verify share validity
BLSPublicKeyVector aggrVerifVector;
for (size_t j = 0; j < N; j++) {
    aggrVerifVector.emplace_back(worker.BuildPubKeyShare(quorum[j].vecP, m.id));
}
CBLSPublicKey aggrPkShare = worker.AggregatePublicKeys(aggrVerifVector);
BOOST_CHECK(m.skShare.GetPublicKey() == aggrPkShare);
```

Reply:
Mhm... Yeah. 
This is redundant, as it is implied by the above `VerifyContributionShare` (the pk, corresponding to the share sent by member `j` to member `m`, is the evaluation point of `j`'s P-vector at `m`'s id, i.e. the output of `BuildPubKeyShare`), and by the aggregation properties of BLS keys (the public key of the secret key, resulting from the aggregation of N secret keys, is equal to the aggregation of the corresponding N public keys).

But still... we can add this test, so we exercise CBLSWorker's `BuildPubKeyShare` and `AggregatePublicKeys` functions too.